Userland/Utilities/pkg/BuildablePort.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "BuildablePort.h"
+#include <AK/LexicalPath.h>
+#include <AK/NonnullRefPtr.h>
+#include <LibCore/File.h>
+#include <LibCore/System.h>
+#include <LibCrypto/BigInt/UnsignedBigInteger.h>
+#include <LibCrypto/Hash/SHA2.h>
+#include <LibFileSystem/FileSystem.h>
+#include <LibLine/Editor.h>
+#include <Shell/AST.h>
+#include <Shell/NodeVisitor.h>
+#include <Shell/Parser.h>
+#include <Shell/PosixParser.h>
+#include <Shell/Shell.h>
+#include <unistd.h>
+
+static ErrorOr<Vector<PortInputFile>> parse_file_list(Span<String> raw_file_list)
+{
+    Vector<PortInputFile> file_list;
+    file_list.ensure_capacity(raw_file_list.size());
+
+    for (auto const& raw_file : raw_file_list) {
+        auto url = URL { raw_file };
+        if (!url.is_valid())
+            return Error::from_string_view("Invalid input file URL"sv);
+
+        auto maybe_fragment = url.fragment();
+        if (!maybe_fragment.has_value())
+            return Error::from_string_view("Input file is missing a hash or git revision"sv);
+        auto fragment = maybe_fragment.release_value();
+
+        if (url.scheme() == "https" || url.scheme() == "http") {
+            auto bigint_hash = Crypto::UnsignedBigInteger::from_base(16, fragment);
+            VERIFY(bigint_hash.length() * Crypto::UnsignedBigInteger::BITS_IN_WORD / 8 == Crypto::Hash::SHA256::digest_size());
+            Crypto::Hash::SHA256::DigestType hash;
+            bigint_hash.export_data({ &hash.data, hash.data_length() });
+            file_list.append({ .url = move(url), .type_specific_info = PortInputFile::HTTPInfo { hash } });
+        } else if (url.scheme() == "git+https") {
+            file_list.append({ .url = move(url), .type_specific_info = PortInputFile::GitInfo { move(fragment) } });
+        } else {
+            return Error::from_string_view("Unsupported input file scheme"sv);
+        }
+    }
```

Comment:
I'd strongly prefer if we did this with the exact same regexes that are used in `.ports_include.sh`.

Reply:
I don't at all see the point of making this code more ugly than it needs to be. The URLs are supposed to be valid URLs, if they're not, any download tool will reject them sooner or later. Using a wrong scheme or a too-early fragment will trip up both our URL parser as well as the regex. Besides: I consider that regex more a hack than anything. Ideally we would use a command-line tool to extract parts of URLs there. If there are any discrepancies between the tools, it's more likely we have a broken regex than a broken URL parser.

Reply:
The point is to make the behavior consistent (and I wouldn't call the proposed code pretty either).

Having two sets of "does regex match? if so, pull out the marked parts" seems a lot cleaner than throwing something that isn't meant to be a URL into our URL parser and then not filtering out the parts that are wrong.